### PR TITLE
Bluetooth: Controller: support FEMs via devicetree

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
The purpose of this pull request is to convert some hardware description in the bluetooth controller subsystem to use devicetree instead of Kconfig. Specifically, radio front end modules (FEMs) are moved to devicetree nodes. The Nordic controller supports this feature. We shouldn't break anything in the openisa controller, but since this controller doesn't have FEM support today, I'm not adding it in.

The first part of the series is prep work:

- update devicetree bindings
- add indirection between existing kconfig options and users in the controller

Then the main patch is `Bluetooth: controller: support FEMs via devicetree`.

We follow that by removing the now-unused Kconfig options.

### CODE REVIEW

**Comment by hubertmis** [comment]
I like this proposal. It's aligned with the direction I described in #30708.
I also like the ""for production" production" approach :)

**Review by anangl — COMMENTED** [review]
LGTM

Nit: the first line of the commit message for the first patch mentions CONFIG_BT_CTLR_GPIO_LNA_PIN instead of CONFIG_BT_CTLR_GPIO_PA_PIN. I actually got confused for a moment when I started reading it. Perhaps it would be good to correct this.

**Comment by carlescufi** [comment]
@czeslawmakarski and @ryanjh please take a look

**Comment by KAGA164** [comment]
I like this proposal to unify the configuration of FEM devices. We can then move nRF21540 EK shield configuration from NCS to upstream Zephyr since there will be use case for it. About CONFIG_BT_CTLR_GPIO_PA and CONFIG_BT_CTLR_GPIO_LNA it depends in my opinion if turning off fem in software make sense, for the nRF21540dk if we don't enable fem then radio won't work since there is no bypass for radio signal but I am not familiar with other fem solution so maybe It is worth to keep possibility to turn off fem in software

**Review by pdunaj — CHANGES_REQUESTED** [review]
nacking to make sure hack is removed.
there seems to be a board that refers to the option being removed:
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/ubx_bmd345eval_nrf52840/ubx_bmd345eval_nrf52840_defconfig#L34

**Review by cvinayak — COMMENTED** [review]
Functional changes to Controller code is nil. Minor explanation requested, otherwise LGTM (not tested myself, hope to do though).

File: subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c
Comment (mbolivar-nordic) [reply subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c]: > Could we have some documentation (internal or any other reference) ...

I've improved the commit logs and comments in the latest draft.

> Specific, `NRF_GPIO_PA_CTLR` what is the purpose?

It's a GPIO controller node identifier. See https://docs.zephyrproject.org/latest/reference/devicetree/api.html#gpio. We need it here to distinguish port 0 from port 1.

> Why, `NRF_GPIO_PA_PSEL` and `NRF_GPIO_PA_PIN` both mean a pin index?

They don't. A PSEL value is not a pin number. See the comments here for an explanation of the difference between the two: https://github.com/zephyrproject-rtos/zephyr/blob/1238410914597aef1e0661e0b2e240a448f2dd60/soc/arm/nordic_nrf/soc_nrf_common.h#L95

Part of the point of this PR is to remove the "port and pin in a single integer" encoding for pins in preference of the standard devicetree `<&gpio_controller pin_number flags>;` type of property.

**Comment by mbolivar-nordic** [comment]
Sorry for the long delay. I've updated this.

With the following app.overlay in the peripheral_hr sample:

```
$ cat app.overlay 
&arduino_spi {
	fem_spi_if: fem-spi@0 {
		label = "FEM_SPI";
		compatible = "nordic,nrf21540-fem-spi";
                spi-max-frequency = <8000000>;
		reg = <0>;
	};
};

/ {
	nrf21540: fem {
		compatible = "nordic,nrf21540-fem";
		tx-en-gpios = <&gpio1 10 0>;
		rx-en-gpios = <&gpio1 11 0>;
		pdn-gpios = <&gpio0 27 GPIO_ACTIVE_LOW>;
		spi-if = <&fem_spi_if>;
	};
};

&radio {
	fem = <&nrf21540>;
};
```

I get the following behavior on nRF52840-DK, which seems reasonable to me:

Screenshot from 2021-11-08 22-01-57

and with:

```
$ cat app.overlay
/ {
	generic: fem {
		compatible = "generic-fem-two-ctrl-pins";
		ctx-gpios = <&gpio1 10 0>;
		ctx-settle-time-us = <5>;
		crx-gpios = <&gpio1 11 0>;
		crx-settle-time-us = <5>;
	};
};

&radio {
	fem = <&generic>;
};
```

I get:

Screenshot from 2021-11-08 22-05-09

I am a bit unsure about this part in radio.c, though:

```c
/* TODO is this OK? */
#define NRF_GPIO_PDN_OFFSET 0
```

If someone (@hubertmis ?) can confirm this is OK, I think this is ready to take out of draft.

**Comment by hubertmis** [comment]
> I am a bit unsure about this part in radio.c, though:
> 
> ```c
> /* TODO is this OK? */
> #define NRF_GPIO_PDN_OFFSET 0
> ```
> 
> If someone (@hubertmis ?) can confirm this is OK, I think this is ready to take out of draft.

It should work, but I would rework it a little. Inside `void radio_gpio_pa_lna_enable(uint32_t trx_us)` there is a `#if` to configure hardware associated with PDN pin only when `FEM_IS_NRF21540`. I would change the condition here to something like `#if defined(HAL_RADIO_FEM_IS_NRF21540) && DT_NODE_HAS_PROP(FEM_NODE, pdn_gpios)`. This way you could remove `#define NRF_GPIO_PDN_OFFSET 0` from `#else` block, because its value would be unused if PDN pin is not configured.

**Comment by mbolivar-nordic** [comment]
>  would change the condition here

Done thanks.

**Comment by mbolivar-nordic** [comment]
> In all FEMs of which I am aware, SoC needs to manage bypass signals. Compiling out software support for FEM if FEM is present in RF path would not work for such devices.

The latest version reflects this comment from @hubertmis .

**Comment by mbolivar-nordic** [comment]
Rebased and added release notes.

**Comment by mbolivar-nordic** [comment]
Addressed a compilation issue on nRF51.

**Comment by carlescufi** [comment]
@anangl @hubertmis @cvinayak could we get a review from you please?

**Comment by carlescufi** [comment]
@brec-u-blox would you mind taking a look, since the only board affected is yours?

**Review by hubertmis — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
I'm aware of the merge conflicts and will fix them when we have enough reviewers.

**Review by carlescufi — APPROVED** [review]
ACK from my side, but do not merge until @cvinayak has approved.

**Review by anangl — APPROVED** [review]
LGTM. Only a few nitpicking comments.

**Comment by brec-u-blox** [comment]
> 
> 
> @brec-u-blox would you mind taking a look, since the only board affected is yours?

LGTM, though the FEM settle time can be reduced to 1uS (Skyworks RFX2411)
image
Tested OK here on a ubx_bmd345eval_nrf52840.

**Review by cvinayak — APPROVED** [review]
As @brec-u-blox has confirmed on hardware, and I do not have any requests from my end, approved. (if anything comes up in the future, when I test/use the feature or any bugs reported, it can be addressed accordingly after this PR is merged).

**Review by nordicjm — APPROVED** [review]

**Comment by cfriedt** [comment]
@mbolivar-nordic - please rebase and resolve conflicts

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic - please rebase and resolve conflicts

Done.

> LGTM, though the FEM settle time can be reduced to 1uS (Skyworks RFX2411)

@brec-u-blox The board files did not specify CONFIG_BT_CTLR_GPIO_PA_OFFSET or CONFIG_BT_CTLR_GPIO_LNA_OFFSET, so they currently (in main) get the default values of 5 for both.

The patch in this PR which moves FEM config to DT is using 5 to try to avoid a change in behavior in a patch that was only meant to provide a change in configuration mechanism.

Since from your report, it's safe to lower the value, I've done so in a separate commit.

**Comment by mbolivar-nordic** [comment]
> LGTM. Only a few nitpicking comments.

Addressed, thanks

**Comment by brec-u-blox** [comment]
@mbolivar-nordic Thanks for the clarification and separate commit.

**Pull request merged**
